File: app/src/main/java/com/example/paetz/yacguide/network/CountryParser.java
package com.example.paetz.yacguide.network;

import com.example.paetz.yacguide.UpdateListener;
import com.example.paetz.yacguide.database.AppDatabase;
import com.example.paetz.yacguide.database.Country;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class CountryParser extends JSONWebParser {

    public CountryParser(AppDatabase db, UpdateListener listener) {
        super(db, listener);
        networkRequests.add(new NetworkRequest(
                NetworkRequestType.DATA_REQUEST_ID,
                baseUrl + "jsonland.php?app=yacguide"));
    }

    @Override
    protected void parseData(NetworkRequestType requestId, String json) throws JSONException {
        // We don't need to care about requestId here; we only have one
        db.countryDao().deleteAll();
        final JSONArray jsonCountries = new JSONArray(json);
        for (int i = 0; i < jsonCountries.length(); i++) {
            final JSONObject jsonCountry = jsonCountries.getJSONObject(i);
            Country c = new Country();
            c.setName(jsonCountry.getString("land"));
            db.countryDao().insert(c);
        }
    }
}

File: app/src/main/java/com/yacgroup/yacguide/network/CountryParser.kt
/*
 * Copyright (C) 2019, 2022 Axel Paetzold
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.yacgroup.yacguide.network

import com.yacgroup.yacguide.database.Country
import com.yacgroup.yacguide.database.DatabaseWrapper

import org.json.JSONArray
import org.json.JSONException
import java.util.*

class CountryParser(private val _db: DatabaseWrapper) : JSONWebParser() {

    private val _countries = mutableListOf<Country>()

    override fun initNetworkRequests() {
        networkRequests = LinkedList(listOf(
                NetworkRequest(
                    NetworkRequestUId(RequestType.COUNTRY_DATA, 0),
                    "${baseUrl}jsonland.php?app=yacguide")
        ))
    }

    @Throws(JSONException::class)
    override fun parseData(requestId: NetworkRequestUId, json: String) {
        val jsonCountries = JSONArray(json)
        for (i in 0 until jsonCountries.length()) {
            val jsonCountry = jsonCountries.getJSONObject(i)
            val country = Country(
                jsonCountry.getString("land")
            )
            _countries.add(country)
        }
    }

    override fun onFinalTaskResolved() {
        _db.deleteCountries()
        _db.addCountries(_countries)
        super.onFinalTaskResolved()
    }
}
